### PR TITLE
Adds notes to Developer Workflow docs

### PR DESCRIPTION
**Pull request by dharmit**
**What this PR does / why we need it**:
What it does: $subject
Why we need it: As a new team member, it took me an embarrassingly long amount of time to make a change in `cmd/hyperconverged-cluster-operator/main.go` and see the changes in `hco-operator` Deployment on OCP repeatably. Hopefully, the notes added here would help the next person do it faster.

**Reviewer Checklist**

> Reviewers are supposed to review the PR for every aspect below one by one. To check an item means the PR is either "OK" or "Not Applicable" in terms of that item. All items are supposed to be checked before merging a PR. 

- [ ] PR Message
- [ ] Commit Messages
- [ ] How to test
- [ ] Unit Tests
- [ ] Functional Tests
- [ ] User Documentation
- [ ] Developer Documentation
- [ ] Upgrade Scenario
- [ ] Uninstallation Scenario
- [ ] Backward Compatibility
- [ ] Troubleshooting Friendly

**Jira Ticket**:

```jira-ticket
https://issues.redhat.com/browse/CNV-27480
```

**Release note**:

```release-note
cluster-sync-olm make target for devs working with an OLM enabled cluster
```

### CODE REVIEW

File: cluster/sync-olm.sh [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+make bundleRegistry
+./hack/build-index-image.sh IMAGE_TAG
+
+# namespace to create catalog source in
+CATSRC_NAMESPACE=""
+CATSRC_NAME="test-hco-catalogsource"
+# namespace to create subscription in
+SUB_NAMESPACE="kubevirt-hyperconverged"
+SUB_NAME="hco-operatorhub"
+CSV_NAME="kubevirt-hyperconverged-operator.v1.10.0"
```
Comment (tiraboschi) [inline cluster/sync-olm.sh:93]: same

File: cluster/sync-olm.sh [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+then
+  create_og
+fi
+
+sub=$(kubectl get -n $SUB_NAMESPACE subscription $SUB_NAME 2>/dev/null)
+if [ -z "$sub" ]
+then
+  # create subscription since it doesn't exist
+  create_subscription
+else
```
Comment (tiraboschi) [inline cluster/sync-olm.sh:127]: same

File: README.md [hunk 3]
```diff
@@ -116,6 +119,16 @@ export CONTAINER_TAG=example
...
 ```
+Modify all the occurrences of `+IMAGE_TO_REPLACE+` in `deploy/index-image/community-kubevirt-hyperconverged/1.9.
+0/manifests/kubevirt-hyperconverged-operator.v1.9.0.clusterserviceversion.yaml` with the container image created by 
+`make container-build` command. Do the same for all the occurrences of `quay.
+io/kubevirt/hyperconverged-cluster-operator:1.9.0-unstable` in the file 
+`deploy/olm-catalog/community-kubevirt-hyperconverged/1.9.0/manifests/kubevirt-hyperconverged-operator.v1.9.0.clusterserviceversion.yaml`.
+
+Build the index image if you would like to use a custom index image to start the Operator:
+```shell
+$ ./hack/build-index-image.sh {ALL|LATEST|<version>} [UNSTABLE]
```
Comment (tiraboschi) [inline README.md:130]: can you please update this?

File: README.md [hunk 4]
```diff
@@ -147,7 +160,7 @@ metadata:
   namespace: openshift-marketplace
 spec:
   sourceType: grpc
-  image: $IMAGE_REGISTRY/$REGISTRY_NAMESPACE/hco-container-registry:$CONTAINER_TAG
+  image: $IMAGE_REGISTRY/$REGISTRY_NAMESPACE/hyperconverged-cluster-index:1.9.0
```
Comment (tiraboschi) [inline README.md:163]: please avoid this

File: cluster/sync-olm.sh [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+  # it's an OCP cluster
+  CATSRC_NAMESPACE="openshift-marketplace"
+fi
+
+catsrc=$(kubectl get -n $CATSRC_NAMESPACE catsrc $CATSRC_NAME 2>/dev/null)
+if [ -z "$catsrc" ]
+then
+  create_catsrc
+else
```
[the remark this answers is not in the document]
Comment (dharmit) [reply cluster/sync-olm.sh:109]: Mainly to be able to reuse stuff. For example, `create_catsrc` and `create_subscription` are repeating within the file.

File: cluster/sync-olm.sh (hunk 2 above)
Comment (dharmit) [reply cluster/sync-olm.sh:127]: https://github.com/kubevirt/hyperconverged-cluster-operator/pull/2286#discussion_r1143388135

File: cluster/sync-olm.sh [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+# - make bundleRegistry
+# - ./hack/build-index-image.sh
+#
+# Prerequisites:
+# - Bunch of variables to build container images and push them to a container registry.
+# - Access to a k8s/OCP cluster that has OLM enabled on it.
+
+# tag for the bundle image
+CONTAINER_TAG="${CONTAINER_TAG:-$(cat /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 8 | head -n 1)}"
+IMAGE_REGISTRY="${IMAGE_REGISTRY:-docker.io}"
```
[the remark this answers is not in the document]
Comment (dharmit) [reply cluster/sync-olm.sh]: Done.

File: cluster/sync-olm.sh [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+function create_catsrc() {
+  cat <<EOF | kubectl create -f -
+      apiVersion: operators.coreos.com/v1alpha1
+      kind: CatalogSource
+      metadata:
+        name: $CATSRC_NAME
+        namespace: $CATSRC_NAMESPACE
+      spec:
+        sourceType: grpc
+        image: $IMAGE_REGISTRY/$REGISTRY_NAMESPACE/hyperconverged-cluster-index:1.10.0
```
[the remark this answers is not in the document]
Comment (dharmit) [reply cluster/sync-olm.sh]: Makes sense. Done.

File: cluster/sync-olm.sh [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+      apiVersion: operators.coreos.com/v1alpha1
+      kind: Subscription
+      metadata:
+          name: $SUB_NAME
+          namespace: $SUB_NAMESPACE
+      spec:
+          source: $CATSRC_NAME
+          sourceNamespace: openshift-marketplace
+          name: community-kubevirt-hyperconverged
+          channel: "1.10.0"
```
[the remark this answers is not in the document]
Comment (dharmit) [reply cluster/sync-olm.sh]: Done.

File: cluster/sync-olm.sh [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+EOF
+}
+
+if [ -z "${REGISTRY_NAMESPACE}" ]; then
+  echo "Please set REGISTRY_NAMESPACE"
+  echo "   REGISTRY_NAMESPACE=rthallisey make cluster-sync-olm"
```
[the remark this answers is not in the document]
Comment (dharmit) [reply cluster/sync-olm.sh]: Done.

File: cluster/sync-olm.sh [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+  echo "   make cluster-sync-olm REGISTRY_NAMESPACE=rthallisey"
```
[the remark this answers is not in the document]
Comment (dharmit) [reply cluster/sync-olm.sh]: It was copy-pasted from `build-registry-bundle.sh`. :)

File: cluster/sync-olm.sh (hunk 1 above)
Comment (dharmit) [reply cluster/sync-olm.sh]: Done.

File: README.md (hunk 4 above)
Comment (dharmit) [reply README.md]: @tiraboschi can you help answer this?

File: cluster/sync-olm.sh [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+if [ -z "$sub" ]
+then
+  # create subscription since it doesn't exist
+  create_subscription
+else
+  # delete the subscription and the CSV, and create sub again
+  kubectl delete -n $SUB_NAMESPACE subscription $SUB_NAME
+  kubectl delete -n $SUB_NAMESPACE csv $CSV_NAME
+  create_subscription
+fi
```
Comment (machadovilaca) [inline cluster/sync-olm.sh:132]: nit: add an empty line at the end of the file here

File: README.md [hunk 12]
```diff
@@ -147,7 +168,7 @@ metadata:
   namespace: openshift-marketplace
 spec:
   sourceType: grpc
-  image: $IMAGE_REGISTRY/$REGISTRY_NAMESPACE/hco-container-registry:$CONTAINER_TAG
+  image: $IMAGE_REGISTRY/$REGISTRY_NAMESPACE/hyperconverged-cluster-index:1.10.0
```
Comment (machadovilaca) [inline README.md:171]: we exported `$CONTAINER_TAG` earlier in the instructions, why have the version here?

File: README.md [hunk 13]
```diff
@@ -100,9 +102,19 @@ $ curl https://raw.githubusercontent.com/kubevirt/hyperconverged-cluster-operato
 If you want to make changes to the HCO, here's how you can test your changes
 through [OLM](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/install/install.md#installing-olm).
 
+The make target `cluster-sync-olm` automates the steps in this section:
+```shell
+$ make cluster-sync-olm
+```
+
+If you would rather want to understand the details, you can find it below.
```
Comment (machadovilaca) [inline README.md:110]: I think this is a bit confusing explaining that the steps below are not necessary
Can you try maybe using subsections?

File: cluster/sync-olm.sh [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+# - make build-webhook
+# - make build-manifests
+# - make container-build
+# - make container-push
+# - Replacing image names in couple of files
+# - make bundleRegistry
+# - ./hack/build-index-image.sh
+#
+# Prerequisites:
+# - Bunch of variables to build container images and push them to a container registry.
```
Comment (machadovilaca) [inline cluster/sync-olm.sh:16]: i would like more that you explicitly listed the variables and their meaning

File: cluster/sync-olm.sh [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+# - Access to a k8s/OCP cluster that has OLM enabled on it.
+
+# tag for the bundle image
+CONTAINER_TAG="${CONTAINER_TAG:-$(cat /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 8 | head -n 1)}"
+IMAGE_REGISTRY="${IMAGE_REGISTRY:-quay.io}"
+# image to be replaced in CSV yaml files
+IMAGE_TO_REPLACE=$IMAGE_REGISTRY/$REGISTRY_NAMESPACE/hyperconverged-cluster-operator:$CONTAINER_TAG
```
Comment (machadovilaca) [inline cluster/sync-olm.sh:23]: this is unused

File: README.md [hunk 16]
```diff
@@ -100,9 +102,19 @@ $ curl https://raw.githubusercontent.com/kubevirt/hyperconverged-cluster-operato
...
+
+If you have made changes to the Operator and/or the webhook part of the code, first run `make build-operator` and/or 
+`make build-webhook`.
+
 Build the HCO container using the Makefile recipes `make container-build` and
 `make container-push` with vars `IMAGE_REGISTRY`, `REGISTRY_NAMESPACE`, and `CONTAINER_TAG`
-to direct it's location.
+to direct its location.
 
 To use the HCO's container, we'll use a registry image to serve metadata to OLM.
```
Comment (machadovilaca) [inline README.md:119]: since this can be used with `make cluster-sync-olm` I suggest moving this section up

File: cluster/sync-olm.sh (hunk 15 above)
Comment (machadovilaca) [reply cluster/sync-olm.sh:23]: did you forget to add the steps as you mentioned above in:

> Modify all the occurrences of `+IMAGE_TO_REPLACE+` in `deploy/index-image/community-kubevirt-hyperconverged/1.10.0/manifests/kubevirt-hyperconverged-operator.v1.10.0.clusterserviceversion.yaml`
> ...

**Review by machadovilaca — CHANGES_REQUESTED** [review]

File: cluster/sync-olm.sh (hunk 15 above)
Comment (dharmit) [reply cluster/sync-olm.sh:23]: It's actually being taken care of by:
```sh
# Image to be used in CSV manifests
HCO_OPERATOR_IMAGE=$IMAGE_REGISTRY/$REGISTRY_NAMESPACE/hyperconverged-cluster-operator
HCO_OPERATOR_IMAGE=$HCO_OPERATOR_IMAGE CSV_VERSION=$CSV_VERSION make build-manifests
```
So the variable should be removed. Thanks for the catch. 👍🏾

File: README.md [hunk 17]
```diff
@@ -116,8 +128,17 @@ export CONTAINER_TAG=example
 # $IMAGE_REGISTRY/$REGISTRY_NAMESPACE/hco-container-registry:$CONTAINER_TAG
 make bundleRegistry
 ```
+Modify all the occurrences of `+IMAGE_TO_REPLACE+` in `deploy/index-image/community-kubevirt-hyperconverged/1.10.0/manifests/kubevirt-hyperconverged-operator.v1.10.0.clusterserviceversion.yaml` with the container image created by
```
[the remark this answers is not in the document]
Comment (dharmit) [reply README.md:131]: Done. I was bitten by this myself.

File: README.md (hunk 13 above)
Comment (dharmit) [reply README.md:110]: By subsections, do you mean something like:
* TL;DR (this has only the one line about using make target)
* Longer version (this should have rest of the details mentioned)

File: README.md (hunk 12 above)
Comment (dharmit) [reply README.md:171]: Corrected.

File: cluster/sync-olm.sh (hunk 11 above)
Comment (dharmit) [reply cluster/sync-olm.sh:132]: Done.

File: cluster/sync-olm.sh (hunk 15 above)
Comment (machadovilaca) [reply cluster/sync-olm.sh]: Then why do we say to:

> Next, modify the occurrences of `+IMAGE_TO_REPLACE+` in `deploy/index-image/community-kubevirt-hyperconverged/1.10.
> 0/manifests/kubevirt-hyperconverged-operator.v1.10.0.clusterserviceversion.yaml` with the container image created by
> `make container-build` command. 
> ....

wouldn't it be easier to say to execute those command too?

File: README.md (hunk 13 above)
Comment (machadovilaca) [reply README.md]: yes, something like that sounds good

File: README.md (hunk 16 above)
Comment (machadovilaca) [reply README.md:117]: just before `make cluster-sync-olm`

**Comment by dharmit** [comment]
Seeing the following failure in [k8s 1.25 tests](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_hyperconverged-cluster-operator/2286/pull-hyperconverged-cluster-operator-e2e-k8s-1.25/1648656727505113088) on this PR makes me wonder if this is a flake because I'm seeing the same failure on #2319 as well (look at this comment https://github.com/kubevirt/hyperconverged-cluster-operator/pull/2319#issuecomment-1514288444):
```
  Run 6/20
  STEP: Creating VMI... @ 04/19/23 12:09:13.491
  STEP: Verifying VMI is running @ 04/19/23 12:09:13.507
  [FAILED] in [It] - /home/prow/go/src/github.com/kubevirt/hyperconverged-cluster-operator/tests/func-tests/node_placement_test.go:123 @ 04/19/23 12:19:13.508
• [FAILED] [651.326 seconds]
[rfe_id:4356][crit:medium][vendor:cnv-qe@redhat.com][level:system]Node Placement validate node placement for vmi [It] [test_id:5679] should create, verify and delete VMIs with correct node placements
/home/prow/go/src/github.com/kubevirt/hyperconverged-cluster-operator/tests/func-tests/node_placement_test.go:117
  [FAILED] Timed out after 600.001s.
  failed to get the vmi Running
  Expected
      <bool>: false
  to be true
  In [It] at: /home/prow/go/src/github.com/kubevirt/hyperconverged-cluster-operator/tests/func-tests/node_placement_test.go:123 @ 04/19/23 12:19:13.508 
```
@tiraboschi @orenc1 @nunnatsa WDYT?

**Comment by machadovilaca** [comment]
@dharmit I like how it looks, please just rebase the commits with the requested changes into the original commits

**Comment by dharmit** [comment]
> @dharmit I like how it looks

@machadovilaca thanks a tonne for your reviews!

> rebase the commits with the requested changes into the original commits

Do you mean to rebase with `main` branch or reduce the number of commits by doing some git-fu? :)

**Comment by machadovilaca** [comment]
@dharmit reduce the number of commits

**Comment by dharmit** [comment]
@machadovilaca I have rebased the last three commits into one using `git rebase -i HEAD~3` and using `fixup` for the two newer commits that I wanted to rebase into the first commit I made after your review. Can you PTAL if I have done what you expected?

If not, can you share an example of what you would rather prefer?

I assumed that the commits will be squashed when merging the PR. What's the process like on this repo? Do we squash or do we merge?

File: Makefile [hunk 18]
```diff
@@ -43,16 +43,16 @@ lint:
 
 build: build-operator build-csv-merger build-webhook
 
-build-operator: $(SOURCES) ## Build binary from source
+build-operator: $(SOURCES) ## Build HCO binary from source from source under cmd/hyperconverged-cluster-operator
```
Comment (tiraboschi) [inline Makefile:46]: I'd remove `from source under cmd/hyperconverged-cluster-operator` since that is only the entry point and it could make more confusion than help

File: Makefile [hunk 19]
```diff
@@ -43,16 +43,16 @@ lint:
...
 	go build -ldflags="${LDFLAGS}" -o _out/hyperconverged-cluster-operator ./cmd/hyperconverged-cluster-operator
 
-build-csv-merger: ## Build binary from source
+build-csv-merger: ## Build csv-merger binary from source under tools/csv-merger/csv-merger.go
 	go build -ldflags="${LDFLAGS}" -o _out/csv-merger tools/csv-merger/csv-merger.go
 
-build-webhook: $(SOURCES) ## Build binary from source
+build-webhook: $(SOURCES) ## Build HCO webhook binary from source under cmd/hyperconverged-cluster-webhook
```
Comment (tiraboschi) [inline Makefile:52]: same

File: Makefile [hunk 20]
```diff
@@ -223,6 +226,8 @@ deploy_hpp:
 validate-no-offensive-lang:
 	./hack/validate-no-offensive-lang.sh
 
+
+
```
Comment (tiraboschi) [inline Makefile:230]: can we avoid this?

File: cluster/sync-olm.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (tiraboschi) [reply cluster/sync-olm.sh:113]: I see a single logical entry of create_catsrc and create_subscription, `kubectl apply` will probably make it more readable

File: cluster/sync-olm.sh [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+if [ -z "${REGISTRY_NAMESPACE}" ]; then
+  echo "Please set REGISTRY_NAMESPACE"
+  echo "   REGISTRY_NAMESPACE=kubevirt make cluster-sync-olm"
+  exit 1
+fi
+
+make build-operator
+make build-webhook
+make container-build
+make container-push
```
Comment (tiraboschi) [inline cluster/sync-olm.sh:40]: Can we make this dependencies in the Makefile?

File: cluster/sync-olm.sh [hunk 22]
```diff
@@ -0,0 +1,146 @@
...
+
+if [ -z "${CSV_VERSION}"]; then
+  CSV_VERSION=latest
+fi
+# Image to be used in CSV manifests
+HCO_OPERATOR_IMAGE=$IMAGE_REGISTRY/$REGISTRY_NAMESPACE/hyperconverged-cluster-operator:$CONTAINER_TAG
+HCO_OPERATOR_IMAGE=$HCO_OPERATOR_IMAGE CSV_VERSION=$CSV_VERSION make build-manifests
+make bundleRegistry
+./hack/build-index-image.sh IMAGE_TAG
```
Comment (tiraboschi) [inline cluster/sync-olm.sh:49]: $CONTAINER_TAG ?

File: cluster/sync-olm.sh [hunk 23]
```diff
@@ -0,0 +1,146 @@
...
+          displayName: KubeVirt HyperConverged
+          publisher: Red Hat
+EOF
+fi
+
+kubectl create namespace kubevirt-hyperconverged 2>/dev/null
+
+og=$(kubectl get -n $SUB_NAMESPACE og $OPERATOR_GROUP_NAME)
+if [ -z "$og" ]
+then
```
Comment (tiraboschi) [inline cluster/sync-olm.sh:104]: we can avoid this logic since we use `oc apply`

File: cluster/sync-olm.sh [hunk 24]
```diff
@@ -0,0 +1,146 @@
...
+          name: $OPERATOR_GROUP_NAME
+          namespace: $SUB_NAMESPACE
+EOF
+fi
+
+sub=$(kubectl get -n $SUB_NAMESPACE subscription $SUB_NAME 2>/dev/null)
+if [ -z "$sub" ]
+then
+  # create subscription since it doesn't exist
+  cat <<EOF | kubectl apply -f -
```
Comment (tiraboschi) [inline cluster/sync-olm.sh:118]: same

File: cluster/sync-olm.sh [hunk 25]
```diff
@@ -0,0 +1,146 @@
...
+        metadata:
+            name: $SUB_NAME
+            namespace: $SUB_NAMESPACE
+        spec:
+            source: $CATSRC_NAME
+            sourceNamespace: openshift-marketplace
+            name: community-kubevirt-hyperconverged
+            channel: $VERSION
+EOF
+fi
```
Comment (tiraboschi) [inline cluster/sync-olm.sh:146]: apply should take care of this

**Review by tiraboschi — CHANGES_REQUESTED** [review]

File: cluster/sync-olm.sh (hunk 22 above)
Comment (tiraboschi) [reply cluster/sync-olm.sh:49]: in this going to work if $IMAGE_REGISTRY/$REGISTRY_NAMESPACE/hyperconverged-cluster-index does not already exist?

Comment (dharmit) [reply cluster/sync-olm.sh:49]: > `$CONTAINER_TAG` ?

I'll admit that I'm confused where exactly `CONTAINER_TAG` makes sense and where `IMAGE_TAG` makes sense. At this particular point, `IMAGE_TAG` seemed more appropriate to me.

> in this going to work if $IMAGE_REGISTRY/$REGISTRY_NAMESPACE/hyperconverged-cluster-index does not already exist?

I don't understand your question. If you are referring to the variables `$IMAGE_REGISTRY`, `$REGISTRY_NAMESPACE` and `$IMAGE_TAG` then I believe  `./hack/build-index-image.sh IMAGE_TAG` particularly depends on `$IMAGE_TAG` and won't work if it's not provided:
```sh
$ /hack/build-index-image.sh
+++ dirname ./hack/build-index-image.sh
++ readlink -e ./hack/../
+ PROJECT_ROOT=/home/dshah/src/hyperconverged-cluster-operator
+ DEPLOY_DIR=/home/dshah/src/hyperconverged-cluster-operator/deploy/olm-catalog
+ OUT_DIR=/home/dshah/src/hyperconverged-cluster-operator/_out
+ PACKAGE_NAME=community-kubevirt-hyperconverged
+ IMAGE_REGISTRY=quay.io
+ REGISTRY_NAMESPACE=dharmit
+ BUNDLE_REGISTRY_IMAGE_NAME=hyperconverged-cluster-bundle
+ INDEX_REGISTRY_IMAGE_NAME=hyperconverged-cluster-index
+ OPM=opm
+ UNSTABLE=
+ [[ 0 -gt 2 ]]
+ [[ 0 -lt 1 ]]
+ help
+ echo 'usage ./hack/build-index-image.sh {ALL|LATEST|<version>} [UNSTABLE]'
usage ./hack/build-index-image.sh {ALL|LATEST|<version>} [UNSTABLE]
+ exit 1
```

Comment (tiraboschi) [reply cluster/sync-olm.sh:44]: 1. `IMAGE_TAG` is a string where the value is `IMAGE_TAG`, if you want to reference a variable named `IMAGE_TAG` you have to use `${IMAGE_TAG}` and the shell parameter expansion will expand it.

2. isn't it going to try to pull `$IMAGE_REGISTRY/$REGISTRY_NAMESPACE/hyperconverged-cluster-index` first?

Comment (dharmit) [reply cluster/sync-olm.sh:44]: > can we automate it somehow?

Yeah, I think so. I'll add it and ping you back.

**Comment by dharmit** [comment]
Closing till FBC work is done.

/close
